### ISSUE TITLE
Enhancement: [no-misused-promises] Add `checkThenables` option (parity with no-floating-promises)

### ISSUE THREAD

**Issue by saberzero1** [typescript-eslint/typescript-eslint#12673]
### Before You File a Proposal Please Confirm You Have Done The Following...

- [x] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22enhancement%3A+plugin+rule+option%22) and found none that match my proposal.
- [x] I have searched the [current rule list](https://typescript-eslint.io/rules/#rules) and found no rules that match my proposal.
- [x] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.
- [x] If I used AI assistance to prepare this issue, I have reviewed and followed the [AI Contribution Policy](https://typescript-eslint.io/contributing/ai-policy/).

### My proposal is suitable for this project

- [x] I believe my proposal would be useful to the broader TypeScript community (meaning it is not a niche proposal).

### Link to the rule's documentation

https://typescript-eslint.io/rules/no-misused-promises/

### Description

`no-floating-promises` gained a `checkThenables` option (default `false`) in #8433 / PR #9263 to avoid false positives on types that structurally match the Thenable interface but are not Promises. Fluent/chaining APIs whose `.then()` method exists for convenience, not for async resolution.

`no-misused-promises` has the same problem but no equivalent escape hatch. Its `checksConditionals` check uses `isAlwaysThenable`, which flags any type whose `.then()` has a callable first parameter, which is the same loose shape that `no-floating-promises` deliberately stopped checking by default.

The concrete case: Obsidian's [BaseComponent](https://docs.obsidian.md/Reference/TypeScript+API/BaseComponent) has a `then` method for fluent chaining:

```ts
export abstract class BaseComponent {
    /** Facilitates chaining */
    then(cb: (component: this) => any): this;
}
```

Every class extending `BaseComponent` (which is most of Obsidian's UI API — `Component`, `Plugin`, `Setting`, views, etc.) gets flagged by `no-misused-promises checksConditionals` in any conditional:

```ts
if (component) { /* ... */ }  // "Expected non-Promise value in a boolean conditional"
```

This is a false positive. The type is not a Promise. `.then()` takes a single callback and returns `this` for chaining. It has no rejection semantics and no second parameter. Notably, the stricter thenable check that `no-floating-promises` uses internally (requiring two callback parameters matching Promise/A+ `resolve`/`reject`) would *not* match this signature. But `no-misused-promises` uses the looser single-parameter check with no way to opt out.

The only current workaround is `checksConditionals: false`, which disables *all* conditional checks.

I'm proposing a `checkThenables` option for `no-misused-promises`, mirroring the one on `no-floating-promises`. When `false` (which I'd suggest as the default, matching `no-floating-promises`), the rule would only flag built-in `Promise` types in conditionals/spreads, not structural Thenables.

This was briefly touched on in https://github.com/typescript-eslint/typescript-eslint/issues/9319#issuecomment-2160017583, where where @bradzacher noted:

> If there's desire we could also add this option to NMP. But you'd be the first request!

Consider this that first request. The same ecosystem pain that motivated `checkThenables` on no-floating-promises (Fastify's fluent `.then()` in #8433) applies equally to `no-misused-promises`. Obsidian's API is another real-world case: a widely-used framework where `.then()` is a chaining method, not a Promise.

[playground reproduction](https://typescript-eslint.io/play/#ts=5.8.3&fileType=.tsx&code=PTAEHkCMGcEsBNYEMB2oBCToFMDCB7AWwAd8VsUAXUACy1CVADpKaKAKASlEO1f3igAZvgBOwgDYBXCtQDGdWCiUBzADSgU%2BagAVRRWDmhMAUCFABZJbEJIJoUdmL74UuZVhlQ%2BIaFbY-AE9iAPhsIWsPMlMlSmxRISQ5AMwcAhIyWVAAbxNQUERoJEgJbHgALlBIfHxS1ABuPNBgACoW0AAxJNgJWEokOOhQBSRrFBVQFuAm-xR2OUhK%2BaJScipK1kNuAF4APgYUQM4NmkNG-JxKABFDYtL4dkK7ssrq2uxUY79T6EaAXxMJjCcgkSEcwzI0HkK0y6wwWDwMLWlEaZjAAEEJPYfN9sDgGODEhJ8aQ4B4AG54mK%2BZYZZHcXL5OSQ95MCT4FTsABEWmojGcBhwXM4-0BzJQUNAAA9QNsIXSsgB%2BUAARlAlQATKiAO6nUqgWmrWQMpqQRxIADW-yAA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6Jge1oFtLlZkiACa1i0Dr0GoMAbXDYciaOOiQANPIVYF2SGQAWiMgGtkAYQ5MhlfJUsBDeNLD44iTdgC%2BHgLrzvnkA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA&tokens=false)

### Fail

```TypeScript
// With checkThenables: true (or current behavior with no option)

interface FluentChainable {
    then(cb: (self: this) => any): this;
}

declare const obj: FluentChainable;

if (obj) {  // Flagged: "Expected non-Promise value in a boolean conditional"
    console.log('not a promise');
}

const x = obj ? 1 : 2;  // Flagged
```

### Pass

```TypeScript
// With checkThenables: false (proposed default)

interface FluentChainable {
    then(cb: (self: this) => any): this;
}

declare const obj: FluentChainable;

if (obj) {  // OK: structural Thenables not checked
    console.log('not a promise');
}

const x = obj ? 1 : 2;  // OK

// Actual Promises are still flagged:
declare const p: Promise<string>;
if (p) {}  // Still flagged (built-in Promise)
```

### Additional Info

- Obsidian's `BaseComponent.then` has existed since [API v0.9.7](https://github.com/obsidianmd/obsidian-api/blob/cc1744324150c632416857c98964f87b1574a5fc/obsidian.d.ts#L503-L508). It's used across the entire Obsidian plugin ecosystem. All `BaseComponent` subclasses (`Component`, `Plugin`, `Setting`, `Modal`, `View`, etc.) inherit this method and trigger the false positive.
- The `no-floating-promises` rule already handled this exact class of problem via `checkThenables: false` (default) in PR #9263, and via a hardened 2-parameter thenable check internally. `no-misused-promises` still uses the loose 1-parameter check with no opt-out.
- An `allowForKnownSafePromises`-style type allowlist (as exists on `no-floating-promises`) would also solve this, but `checkThenables` is the more general solution and has direct precedent.

**Comment by saberzero1** [comment]
I'm a [community reviewer](https://obsidian.md/help/credits#Community+review) for Obsidian. It seems the recent batch PRs sent out to help theme authors prepare for the [1.13 breaking changes](https://obsidian.md/changelog/2026-07-30-desktop-v1.13.4/#:~:text=The%20%2D%2Dcallout%2Dcolor%20variable%20now%20expects%20a%20valid%20CSS%20color%2E%20This%20is%20a%20breaking%20change) to `--callout-color`/`--canvas-color` flags my account for your automation detection tooling.

**Comment by hesprs** [comment]
> Activity patterns show signs of automation.

Same request here. I boycott this absurd scan.

**Comment by bradzacher** [comment]
The reason this is the first request in a decade for this functionality is because the only way to hit it is if you're writing useless conditions.

```ts
declare const obj: FluentChainable;

if (obj) {  // Flagged: "Expected non-Promise value in a boolean conditional"
    console.log('not a promise');
}
```

This condition is useless and constant - `obj` is never falsey - which is why the rule looks at this and reports - because it assumes as a constantly truthy thenable you probably meant instead that you meant to resolve the thenable to its value.

If you can provide an example that isn't a useless/constant condition then we can discuss further.

**Comment by saberzero1** [comment]
Fair point on `checksConditionals`. A non-nullable thenable in a conditional is indeed a constant/useless condition regardless.

However, the false positive also affects `checksVoidReturn`. The Obsidian API's `Setting` class has a `.then(cb)` method for fluent chaining, so any code that implicitly returns a `Setting` into a void-returning context gets flagged:

```typescript
// Minimal reproduction of Obsidian's Setting class (fluent API with .then() for chaining)
// https://github.com/obsidianmd/obsidian-api/blob/cc1744324150c632416857c98964f87b1574a5fc/obsidian.d.ts#L5849-L5854
declare class Setting {
    constructor(containerEl: HTMLElement);
    setName(name: string): this;
    setDesc(desc: string): this;
    addButton(cb: (button: { setButtonText(text: string): void }) => void): this;
    /** Facilitates chaining */
    then(cb: (setting: this) => any): this;
}

declare const containerEl: HTMLElement;
const items = ['item1', 'item2', 'item3'];

// Expression-body arrow implicitly returns Setting to forEach's void callback
items.forEach((item) =>
    new Setting(containerEl)
        .setName(item)
        .addButton((button) =>
            button.setButtonText('Remove')
        )
); // "Promise returned in function argument where a void return was expected"
```


[playground reproduction](https://typescript-eslint.io/play/#ts=5.8.3&fileType=.tsx&code=PTAEFkEsDtIWwIYBtQCcCmAHVB7AJgK4DGALpDtKDgGagDyARgM6R6QLQDkToAyuiTLQA5qCJIETHgApqSAumglQAQQAKASVAB3SCQAWoAHQHF0gJShqOVGP0IYMYeYBQedOIQYxEqXwFCogDeLqBiFEwkqMQkNtJEFCQO0OioAKJIAFygABIAKuAAMhnocIok5gDcoaBMAgByCGXS0E3o2ZGoTqAAPqAAIjhEBGVKAGKoCMKjFdkGkEzVYXUk-ehMRNLuGx1R3X2DwzMTUzPmc-oLS6AIeHgAQgSCFPEM2fE4cJgU5dmPz9AAMKfb4pJSWAC8AD4btAAJ7nUDzRY1YAAKjRoDGCCIkCQegQJHWdmS3TRwBqpmgr3eK0CFwWkJhHARDJRAF8XG4PBJvJ4-P9YkCQT8lKAQssBIKKHl0AAPEjSIkK3ZdESI5HVTnczx8iLKBJKZKpDLZfJFEozaqGyKgPSlHgQ0AAbU49rgAEZOAAaUBuolwABMPr97oAzJwALrVFwgUBpOXYdYsCgAWgY%2BDhN1QuG0dq%2B%2BNxJCQWYwJAIqGgPH4gm6sSsNjSOP03FAADccKwxMgkAwcQBrFzuphGazpZvSaTupk1FJ5muBD5GmAmpCuMJhIwrRrNac1Te3B5PIWThjHigzjcbs8ArdS8-QWUK6ScABKpRwbfQnHXG9cVRcIA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6Jge1oFtLlZkiACa1i0Dr0GoMkRNHHRI4MAF8QKoA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA&tokens=false)

This isn't a useless condition; it's a standard fluent-chaining pattern where the return value is irrelevant. The `Setting` is not a Promise; `.then()` is a chaining convenience method. `checksSpreads` is similarly affected (`{ ...setting }` gets flagged), though that's a rare pattern in practice.

**Comment by hesprs** [comment]
> The reason this is the first request in a decade for this functionality is because the only way to hit it is if you're writing useless conditions.

Please do not judge the "usefulness" of a code simply from a minimal repro. The lint rule can break easily in production codebase.

For example, [TypeScript has known limitations in control flow analysis](https://github.com/microsoft/TypeScript/issues/9998). The conditional statements shown below are valid guards, but the lint rule just shows false positive:

```TypeScript
type BaseComponent = { then(cb: () => any): unknown }

declare const existingComponent: BaseComponent;
declare const needNullify: boolean;
declare let component: BaseComponent | undefined;

function doSomething() {
  if (needNullify) {
    console.log('This is complex logic');
    component = undefined;
  } else component = existingComponent;
}

if (!component) throw new Error();

doSomething();

// Below are false positives
const x = component ? 1 : 2;

while (component) {
  break;
}
```

[Playground link](https://typescript-eslint.io/play/#ts=6.0.2&fileType=.tsx&code=C4TwDgpgBAQghgZwgYQPYFsyoHYW8KAXigG8pgALPACgGMAjALimoEoiA%2BKObEV5gK7YA1tlQB3bFAC%2BAKFkATCLQA2cAE7RaOBAQgAPAJa7D2AOZpMOPMGbwklrLnwBuRcrWao27Lqi4IBQA5ARUVQwAzEGZ6VFQVCB43JVUNaASCbStnW1hEFAwnGygAHyghJQjTQLdZCKFaYEMcKAVUAGUMCEpTMzZSWSgoSJYA4NDwqPYSQaHvHXiIADoVVD6AcgAVCmNhhHnMBP0oVbNDWnXWNzmDovwicuxK6oVrmSgIFSRb63viA2MTXMjl%2BwDcclkI2oAEIsndgOxKOoJP4IOIoABRdTI9RsWptTrobo7cx4%2BQAenJsE%2BKLSUAicC%2B0CwCEMTQAbhAELIfH5jsQ4aCoAB%2BKAARigzAATLVxDsEixBTlprN6Jo4MJwUA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6Jge1oFtLlZkiACa1i0Dr0GoMAbXDYciaOOiQANPIVYF2SGQAWiMgGtkAYQ5MhlfJUsBDeNLD44iTdgC%2BHgLrzvnkA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA&tokens=false)
